src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryRegEx.cs

```diff
@@ -0,0 +1,343 @@
+
+                    if (!int.TryParse(groupName, out _))
+                    {
+                        if (!fields.ContainsKey(validName))
+                        {
+                            fields.Add(validName, new NamedValue(validName, StringValue.New(g.Value)));
+                        }
+                        else
+                        {
+                            fields[validName] = new NamedValue(validName, StringValue.New(g.Value));
+                        }
+                    }
+
+                    if (i > 0)
+                    {
+                        subMatches.Add(g.Value);
+                    }
+                }
+
+                fields.Add(SUBMATCHES, new NamedValue(SUBMATCHES, TableValue.NewSingleColumnTable(subMatches.Select(s => StringValue.New(s)).ToArray())));
+
+                return RecordValue.NewRecordFromFields(fields.Values);
+            }
+        }
+
+        internal abstract class MatchFunctionBase : BuiltinFunction, IAsyncTexlFunction
+        {
+            public override bool IsSelfContained => true;
+
+            public string DefaultMatchOption { get; }
+
+            public TimeSpan RegExTimeout { get; }
+
+            public MatchFunctionBase(TimeSpan regExTimeout, string name, DType returnType, StringGetter description, string defaultMatchOption)
+                : base(name, description, FunctionCategories.Text, returnType, 0x0, 2, 3, DType.String, DType.String, DType.String)
+            {
+                RegExTimeout = ValidateTimeout(regExTimeout);
+                DefaultMatchOption = defaultMatchOption;
+            }
+
+            public override IEnumerable<StringGetter[]> GetSignatures()
+            {
+                yield return new[] { TexlStrings.MatchArg1, TexlStrings.MatchArg2 };
+                yield return new[] { TexlStrings.MatchArg1, TexlStrings.MatchArg2, TexlStrings.MatchOptions };
+            }
+
+            public override bool CheckTypes(CheckTypesContext context, TexlNode[] args, DType[] argTypes, IErrorContainer errors, out DType returnType, out Dictionary<TexlNode, DType> nodeToCoercedTypeMap)
+            {
+                bool fValid = base.CheckTypes(context, args, argTypes, errors, out returnType, out nodeToCoercedTypeMap);
+
+                if (argTypes[0].Kind != DKind.String && argTypes[0].Kind != DKind.ObjNull)
+                {
+                    fValid = false;
+                    errors.EnsureError(DocumentErrorSeverity.Severe, args[0], TexlStrings.ErrStringExpected);
+                }
+
+                if (argTypes[1].Kind != DKind.String && argTypes[1].Kind != DKind.OptionSetValue)
+                {
+                    fValid = false;
+                    errors.EnsureError(DocumentErrorSeverity.Severe, args[0], TexlStrings.ErrStringExpected);
+                }
+
+                if (argTypes.Length == 3)
+                {
+                    if (argTypes[0].Kind != DKind.String && argTypes[0].Kind != DKind.ObjNull)
```

Comment:
>if (argTypes[0].Kind != DKind.String && argTypes[0].Kind != DKind.ObjNull) [](http://example.com/codeflow?start=20&length=74)

Wouldn't this be already caught by line 227?

Reply:
Maybe you meant to compare index 3, but I think that may also be caught by base.CheckType at line 225

Reply:
All this code has been changed